### DISCUSSION TITLE
Assigning to readonly from local functions in constructors

### DISCUSSION THREAD

**Discussion by oocx** [dotnet/csharplang#5418] [Language Ideas] [answered]
If I have a constructor or static constructor that performs multiple different initialization tasks, I'd like to be able to move them to local functions just to group related code together.

However, in a local function, I'm not allowed to assign to readonly fields.

Shouldn't this be possible? The local function is called from the constructor and (as far as I know) can never be called outside of the constructor, so why can't it assign to readonly fields?

```csharp
class Demo
{

    Demo()
    {
        void MyLocal()
        {
            foo = "bar"; // this is not allowed!
        }

        MyLocal();
    }

    readonly string foo;
}
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
There are two things at play.  First, the local function is compiled down to a real method.  So it would be necessary to allow real methods to be able to assign here without runnign afoul of any rules in places at the runtime level.

Second, while in the above example you can examine and see things are safe, consider that a local function can be captured as a delegate and then called outside of hte constructor (which would not be safe).  As such, we'd need a concept of local functions that became illegal upon capture (something we have no concept of today).

**Reply by oocx** [reply]
Thank you for the explanation! I already thought there are probably things like this that I did not think of. Now it's clear why this doesn't work.

**Comment by Joe4evr** [comment]
#1873

Of note, [it *does* work if you pass the field by `ref`](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWAChiACAEVTAHtKBvSy6zuhxgCgEoOXNhS5jqBACzUAsjgAyjAMYQANr3SosEkgAZqAZ0GjxnEadMHqAXmoAiYBHR2A3ENMBfdidNzFK9U1tLEZGfjcfai8Kd2pNCAATRhhVHB19EMYIjyA===).

**Reply by oocx** [reply]
Thanks! I assumed there must already be an issue for this, but was unable to find it.
